exercises/practice/perceptron/runtests.jl

```diff
@@ -0,0 +1,86 @@
+using Test, Random
+include("perceptron.jl")
+
+function runtestset()
```

Comment:
I think we should include some tests with manually specified input data of just a few points to make this more approachable and also as good practice (e.g. first few tests could be spaces with just 2-4 points placed manually).

Reply:
Sorry, I'm not sure if I've understood, but there are four tests with (six) manually specified points to illustrate a couple of different possible orientations of a hyperplane (testset "Low population"). After that the 40 pseudorandomly generated tests begin (testset "Increasing Populations"). Was this what you meant?

Reply:
I meant we could take some of the manually specified examples out of the runtestset() function and test them without the support functions, just so that it's less mystifying to a student reading the tests.

And maybe we should have the student write their own function for finding the computed label of a point?

e.g.

```jl
# Student must implement both `perceptron(points, labels)` and `classify(boundary, point)`

# Originally provided points should be classified correctly
boundary = perceptron([[0,0], [3, 3]], [1, -1])
@test classify(boundary, [0, 0]) == 1
@test classify(boundary, [3, 3]) == -1

# Adding more points constrains the location of the boundary so that we can test
# the classification of unseen points.
boundary = perceptron([[0,0], [1, 0], [0, 1]], [-1, 1, 1])
@test classify(boundary, [0, 0]) == -1
@test classify(boundary, [2, 5]) == 1
```

Reply:
Sure! I can throw a couple in there like your first examples, but I've got a question about the `classify`:
Due to the wide range of possible decision boundaries returned by Perceptron, beyond carefully selected examples, I'm not sure if testing classification of unseen points is viable. Also, since the algo is effectively just the three parts of classify + update + repeat, couldn't requiring a separate classification function introduce an unnecessary constraint on the task?

Reply:
Cool!

I don't see why it would constrain the student or the task for us to ask them to provide a classify function. What do you mean by that?

I agree that testing with unseen points is a bit tricky, but it's not super hard given that we control the input data and it's not too bad for us to either only include hand-written tests or to do some geometry to ensure our unseen pointd will always be classified correctly by a valid linear boundary.

Reply:
I guess my thoughts were that, since classification is already an integral part of the algorithm, requiring a separate classification function doesn't really add anything substantive to the task. One thing this could constrain is that it could encourage a modular implementation over a unitary one, as opposed to leaving this option more open to the student. This is not a big issue, but I don't really see the benefit of adding the extra requirement since it's already built in.

Reply:
I think a good reason is that it helps explain how these things work. A real supervised learning interface would usually have a function for training the classifier and another for using the output of the training to do some classification.

Reply:
I think keeping it at functions might be better, so that there's only one confusing thing to work out at a time:

```jl
fit(points, labels) -> boundary
predict(boundary, point) -> label

# or possibly
predict(boundary, points) -> labels
```

If we go with a struct, then I think we'd end up with something like this if we followed what's done in MLJ and Lux.jl:

```jl
struct Perceptron
  # parameters for customising the perceptron go here, e.g. learning rate r
  r::Float64
  Perceptron(; r) = new(r)
end

struct Model
  w::Vector{Float64}
  # ...
end

model = fit(Perceptron(r=0.2), points, labels)
predicted_labels = predict(model, unseen_points)
```

Which is maybe a bit too complicated?

Reply:
Since the only non-trivial hyperparameter that the basic Perceptron algo has is the initial hyperplane, I think we could get away with pre-defining a single struct and leaving the student to define the `fit` and `predict` methods. Maybe making the slug something like:

```julia
mutable struct Perceptron
    # instantiates Perceptron with a decision boundary (this struct can remain unmodified)
    dbound::Vector{Float64}
    Perceptron() = new([0., 0., 0.])
end

function fit!(model::Perceptron, points, labels)::Vector{Float64}
    # updates and returns the decision boundary: model.dbound

end

function predict(model::Perceptron, points::Vector{Vector{Float64}})::Vector{Float64}
    # returns a vector of the predicted labels of points against the model's decision boundary

end
```
This would leave only methods to be defined while still giving a look into how structs work and a more realistic interface (whose use will be able to be seen in the tests).

On whether the `predict` method should be for a single point or multiple, I would say that multiple is better, since it's easier to wrap a single point in a Vector as input for `predict` rather than to ask the student to handle this difference inside the method (not to mention it's closer to what they'd see in practice).

Reply:
Is the learning rate trivial?

I'm fine with either of these approaches (two methods, with or without a struct). Are you happy to make the changes to the PR?

Reply:
>Is the learning rate trivial?

Sorry, over-generalization from me there :) I guess I'm considering the basic (pedagogical) Perceptron, i.e. one provided with small, dense, separable populations, since, under separability, the learning rate doesn't affect either the final loss or the upper bound of the number of errors the algorithm can make. That said, I was wrong to say the initial hyperplane affects these, since it doesn't either. It just seems non-trivial to me because it can be returned.

I'll try to make the necessary changes to the PR today and/or tomorrow :)

Reply:
>I agree that testing with unseen points is a bit tricky, but it's not super hard given that we control the input data and it's not too bad for us to either only include hand-written tests or to do some geometry to ensure our unseen pointd will always be classified correctly by a valid linear boundary.

I was thinking about this, and it appears it is redundant to test unseen points which are guaranteed to be on the correct side of any possible decision boundary for a population. That's because there are two hyperplanes which define the limits of a cone of possibility, and each of these limits can be defined by two points, one from each class. These two points work like support vectors, and are robust to any additional (guaranteed valid) unseen points. This means only the support vectors need to be classified correctly to verify that the hyperplane is a valid decision boundary, and this is already done under the current testing.

For this reason I think it would be best to leave out testing of unseen points, since, for students encountering this for the first time, it could give the unintended impression that Perceptron will always correctly classify unseen points.

Any thoughts?

Reply:
Is there anything you would add (e.g. subtyping) to my suggestion for a slug, or do you think I could just copy and paste?

Reply:
I don't think the struct is really adding much, so I'd remove it. You can pick what you like, tho.

If I try the exercise and hate the struct then I might veto it, but not until then.

Reply:
> I think we should include some tests with manually specified input data of just a few points to make this more approachable and also as good practice (e.g. first few tests could be spaces with just 2-4 points placed manually).

Agree with this

Reply:
> Agree with this

Could you be more specific? There are already four tests with six manually specified points which check for different possible orientations of a decision boundary. 

Beyond this, we've had an extensive conversation about using "unseen points". (TLDR: I believe testing unseen points to be potentially more detrimental to understanding than beneficial)

The other ideas for tests were of the type to see if the student is returning the correct object (vector of three real numbers), etc. Is there something else you were thinking?